doc/api/async_hooks.md

```diff
@@ -0,0 +1,677 @@
+# Async Hooks
+
+> Stability: 1 - Experimental
+
+The `async-hooks` module provides an API to register callbacks tracking the
+lifetime of asynchronous resources created inside a Node.js application.
+It can be accessed using `require('async-hooks')`.
+
+## Terminology
+
+An async resource represents either a _handle_ or a _request_.
+
+Handles are a reference to a system resource. Some resources are a simple
+identifier. For example, file system handles are represented by a file
+descriptor. Other handles are represented by libuv as a platform abstracted
+struct, e.g. `uv_tcp_t`. Each handle can be continually reused to access and
+operate on the referenced resource.
+
+Requests are short lived data structures created to accomplish one task. The
+callback for a request should always and only ever fire one time. Which is when
+the assigned task has either completed or encountered an error. Requests are
+used by handles to perform tasks such as accepting a new connection or
+writing data to disk.
+
+## Public API
+
+### Overview
+
+Following is a simple overview of the public API. All of this API is explained in
+more detail further down.
+
+```js
+const async_hooks = require('async_hooks');
+
+// Return the id of the current execution context.
+const cid = async_hooks.currentId();
+
+// Return the id of the handle responsible for triggering the callback of the
+// current execution scope to fire.
+const tid = async_hooks.triggerId();
+
+// Create a new AsyncHook instance. All of these callbacks are optional.
+const asyncHook = async_hooks.createHook({ init, before, after, destroy });
+
+// Allow callbacks of this AsyncHook instance to fire. This is not an implicit
+// action after running the constructor, and must be explicitly run to begin
+// executing callbacks.
+asyncHook.enable();
+
+// Disable listening for new asynchronous events.
+asyncHook.disable();
+
+//
+// The following are the callbacks that can be passed to createHook().
+//
+
+// init() is called during object construction. The resource may not have
+// completed construction when this callback runs, therefore all fields of the
+// resource referenced by "id" may not have been populated.
+function init(id, type, triggerId, resource) { }
+
+// before() is called just before the resource's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWrap), and should be called exactly 1
```

Comment:
Reason I said "should be" was to communicate at the time that since it's an experimental API there's a possibility that it may not be called, and if it isn't then there's a bug.